src/tribler-core/tribler_core/modules/metadata_store/community/tests/test_gigachannel_community.py

```diff
@@ -289,3 +290,51 @@ async def test_drop_silent_peer_from_channels_map(self):
             self.nodes[1].overlay.send_remote_select(self.nodes[0].my_peer, **kwargs_dict)
             await self.deliver_messages(timeout=1)
             self.nodes[1].overlay.channels_peers.remove_peer.assert_not_called()
+
+    @pytest.mark.timeout(0)
+    async def test_remote_select_channel_contents(self):
```

Comment:
I noticed this test in passing: this test does not follow the [AAA pattern](https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/) and does not make use of the available test case code.

The pattern here is _Arrange-Act-Assert-Act-Assert-Act-Assert-Act-Assert_. Now I'm not for anally following all software patterns to a T, but your pattern puts a lot of stress on a reviewer and makes it harder to interpret the unit test suite's failing tests. Splitting this test into four separate tests would adhere to the Arrange-Act-Assert standard and I see no technical reason to deviate from the standard in this case.

Of secondary concern is the code duplication in this test file. There is high duplication in the arranging of this test with `test_remote_search_mapped_peers` for example. This can be refactored into a common method of `TestGigaChannelUnits`.
The volume and complexity of this test can be further reduced by using the available shortcuts. For example, `self.nodes[0].overlay.mds.TorrentMetadata` can be replaced with `TestGigaChannelUnits.torrent_metadata(0)`. Another commonly used shortcut you can employ is `self.overlay(i)` instead of `self.nodes[i].overlay`.

Reply:
I refactored the offending test in 3 separate tests. However, I still use the `TestBase` thing, and I don't understand how to combine that with `pytest` fixtures. I'm not aware of a way to apply `Patch` context managers to a whole test w/o pytest stuff. 

We should refactor the whole `TestBase` into a pytest fixture eventually.

Reply:
Nice 👍 this looks 100x better now.

> We should refactor the whole `TestBase` into a pytest fixture eventually.

Why? Fixtures are not some magical construct that solve.. well.. _any_ problems at all in this case. In fact, they make your life actively worse.

1. To stage communities you need to keep a state. This state is being kept explicitly in `TestBase` superclass now. It is also possible to keep this state in a fixture, but you would still need the a `TestBase` **fixture** to juggle the test state anyway. So, you're not lowering the number of objects in your stack, that's staying the same.

2. As a downside for keeping state in fixtures, tests no longer explicitly point to the `TestBase` fixture (e.g., `from ipv8.test.base import TestBase`) and programmers will have to go on an archeological adventure to find out what is providing the network mocking functionality. Trying to find where your internals are defined costs time. Trying to figure out what fixtures your fixtures are using takes even longer.

3. In the `TestBase` you can easily overwrite functionality in a subclass you fit your specific test case. Using fixtures you have to redefine the fixtures in a smaller scope (e.g., in the session scope instead of the global scope) and hope and pray you caught all the side effects. If you had used a normal `TestCase` subclass your IDE shows you exactly what objects are maintained where. All the code is grouped together and you can easily what is going on.

4. In a normal class you can dynamically change your data. You can, in the middle of a unit test, modify the attributes of `TestBase` (e.g., add nodes and whatnot). You cannot (easily) modify a fixture that is not a manager (adding to a fixture that is a list or a dict or something is fine, replacing an immutable fixture at runtime is not going to happen easily).

5. You're losing compatibility. Subclasses of `TestCase`, like `TestBase`, can be run on pretty much every test runner for example `nose`, `unittest` and `pytest`. On the other hand `pytest` syntax doesn't work without.. well.. `pytest`.

6. It costs time and effort to convert `TestBase` into a fixture and will most likely, like all refactorings, bring with it some bugs.

The only upside to converting `TestBase` to a fixture I can see is uniformity of all tests in Tribler. However if uniformity is what you want, for the reasons I presented above, I'd favor converting everything back to `TestBase` instead of converting everything to `pytest`. The original reason to switch to `pytest` was because it would magically fix a hiddenservices test bug, [which it didn't](https://github.com/Tribler/tribler/issues/5582).

To end with some conventional wisdom: if it ain't broke, don't fix it. `pytest` works perfectly fine with the existing `TestBase`. Or, to quote [C4](https://hintjens.gitbooks.io/social-architecture/content/chapter4.html):

> A patch SHOULD be a minimal and accurate answer to exactly one identified and agreed problem.

> The user or Contributor SHOULD seek consensus on the accuracy of their observation, and the value of solving the problem.

I don't see a problem being solved nor do I see the added value of the solution.

Reply:
The point of fixtures is to arbitrarily combine external states for individual test cases, mixin-style. It is not easy to achieve the same level of code reuse by subclassing `TestBase`. Also, I do not agree with your point (2). It was very daunting for me to delve deep into the bowels of `TestBase` classes hierarchy with all the non-relevant stuff around. Conversely, it is very easy for me every time to follow the fixtures.

Anyways, the ability to `yield` from a fixture is amazing, especially in combination with context managers. You should try it.

Reply:
> The point of fixtures is to arbitrarily combine external states for individual test cases, mixin-style. It is not easy to achieve the same level of code reuse by subclassing TestBase.

Sure. I agree that in some cases a mix-in pattern can be useful. However, you can add a mix-in to a class, you cannot add a superclass to a mix-in.

Also, "combining external states" easily is a minefield. Mix-ins make your code more and more complex as your software grows. Here's a breakdown by a React developer if you want that backed-up by developers of large frameworks: https://medium.com/@dan_abramov/mixins-are-dead-long-live-higher-order-components-94a0d2f9e750

> It was very daunting for me to delve deep into the bowels of TestBase classes hierarchy with all the non-relevant stuff around. Conversely, it is very easy for me every time to follow the fixtures.

Is it though? Let's take an example, suppose we want to know where `seed_config` comes from in this file:

https://github.com/Tribler/tribler/blob/089fb6ab1c597b90a051f35350d0e3703d3a07a6/src/tribler-core/tribler_core/modules/tunnel/test_full_session/test_tunnel_community.py#L68

It's not defined in this file. It's not even imported in this file. **First full project namespace search.** Aha, it's coming from here:

https://github.com/Tribler/tribler/blob/041a26f3d8c0cdd483f369e753e6482036a905d0/src/tribler-core/tribler_core/conftest.py#L71

That is created through `tribler_config` (thankfully defined in the same file) and `tmpdir_factory`. **Second full project namespace search.** Ah, not defined in this project. **Pytest namespace search.** Aha! There it is, finally we know how this fixture is created.

This is not easy at all, I had to do three full searches manually in two different scopes. I was also lucky there were no name or scope collisions in this case. In my IDE and even with GitHub's Python integration, this would've taken three clicks. The problem is the same as with star-imports, you get a bucketload of stuff with side-effects with no clue what the source is. You say classes include more "non-relevant stuff", I've both shown and experienced the opposite.

> Anyways, the ability to yield from a fixture is amazing, especially in combination with context managers. You should try it.

I have tried it and it, once again, has its uses. However, sometimes I require fine-grained control of the setup and shutdown of components in different fixtures. In those cases, fixtures become a mess. The same point holds as before, you can do the exact same thing in a class, but you can't bring the fine-grained control into a fixture (without turning your code into fixture spaghetti).

Reply:
> I had to do three full searches manually in two different scopes. 

For me, it was a matter of two Ctrl-LMB clicks. Or two `gd` commands from the keyboard :shrug: (VIM emulation) What IDE do you use?

>you can't bring the fine-grained control into a fixture

Can't we just have the best of both worlds? Use `TestBase` and add a fixture here and there somehow? Or at least add `yield`-base pseudo-fixtures to `TestBase`?

Reply:
> What IDE do you use?

I use PyCharm, like most of our contributors, and I use GitHub to review PRs.

> Can't we just have the best of both worlds? Use TestBase and add a fixture here and there somehow? Or at least add yield-base pseudo-fixtures to TestBase?

Yes and that is all I'm bargaining for. Just use `TestBase` subclasses where appropriate and use fixtures where appropriate. Don't force fixtures where they are not needed or they are actively harmful. It is perfectly reasonable to provide fixture shortcuts for `TestBase` for program flows that could benefit from it. However, as you first proposed, I consider it harmful to straight up refactor `TestBase` into a fixture.